python/ray/data/dataset.py

```diff
@@ -498,46 +510,43 @@ def map_batches(
         context = DatasetContext.get_current()
 
         def transform(
-            block: Block,
+            blocks: Iterable[Block],
             batch_fn: BatchUDF,
             *fn_args,
             **fn_kwargs,
         ) -> Iterable[Block]:
             DatasetContext._set_current(context)
             output_buffer = BlockOutputBuffer(None, context.target_max_block_size)
-            block = BlockAccessor.for_block(block)
-            total_rows = block.num_rows()
-            max_batch_size = batch_size
-            if max_batch_size is None:
-                max_batch_size = max(total_rows, 1)
-
-            for start in range(0, total_rows, max_batch_size):
-                # Build a block for each batch.
-                end = min(total_rows, start + max_batch_size)
-                # Make sure to copy if slicing to avoid the Arrow serialization
-                # bug where we include the entire base view on serialization.
-                view = block.slice(start, end, copy=batch_size is not None)
```

Comment:
> To be clear, the current hard-coded ensure_copy=True is preserving the existing semantics, i.e. this is the status quo in master; exposing this to the user would be providing the option to the user to change these semantics, to get a zero-copy, read-only batch provided to their UDF.

I may be miss something, but if `batch_size=None`, previously we don't copy the block here, and currently we copy the block in `batcher.py:L109`?

Reply:
Ah, good point, those semantics should be preserved, let me make sure that's the case. Good catch!

Reply:
Hmm actually, the batch mutating test that I added fails for the `batch_size=None` case in both master and in this PR. I'll remove that test case since (1) this is current semantics in master,  and (2) we can consider `batch_size=None` to be a zero-copy fast path.